src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java

```diff
@@ -1984,213 +1979,255 @@ public void fetchMetricsBasedOnProfileAndDatasource(KruizeObject kruizeObject, T
+            SimpleDateFormat sdf = new SimpleDateFormat(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT, Locale.ROOT);
+            // Create the client
+            GenericRestApiClient client = new GenericRestApiClient(dataSourceInfo);
 
             Double measurementDurationMinutesInDouble = kruizeObject.getTrial_settings().getMeasurement_durationMinutes_inDouble();
             List<K8sObject> kubernetes_objects = kruizeObject.getKubernetes_objects();
 
-            // Iterate over Kubernetes objects
             for (K8sObject k8sObject : kubernetes_objects) {
                 String namespace = k8sObject.getNamespace();
-                String workload = k8sObject.getName();
-                String workload_type = k8sObject.getType();
-                HashMap<String, ContainerData> containerDataMap = k8sObject.getContainerDataMap();
-                // check if containerDataMap is not empty
-                if (!containerDataMap.isEmpty()) {
-                    // Iterate over containers
-                    for (Map.Entry<String, ContainerData> entry : containerDataMap.entrySet()) {
-                        ContainerData containerData = entry.getValue();
-                        String containerName = containerData.getContainer_name();
-                        if (null == interval_end_time) {
-                            LOGGER.info(KruizeConstants.APIMessages.CONTAINER_USAGE_INFO);
-                            String queryToEncode = null;
-                            if (null == maxDateQuery || maxDateQuery.isEmpty()) {
-                                throw new NullPointerException("maxDate query cannot be empty or null");
-                            }
-
+                // fetch namespace related metrics if containerDataMap is empty
+                NamespaceData namespaceData = k8sObject.getNamespaceData();
+                // determine the max date query for namespace
+                String namespaceMaxDateQuery = maxDateQuery.replace(AnalyzerConstants.NAMESPACE_VARIABLE, namespace);
+
+                if (null == interval_end_time) {
+                    LOGGER.info(KruizeConstants.APIMessages.NAMESPACE_USAGE_INFO);
+                    String dateMetricsUrl = String.format(KruizeConstants.DataSourceConstants.DATE_ENDPOINT_WITH_QUERY,
+                            dataSourceInfo.getUrl(),
+                            URLEncoder.encode(namespaceMaxDateQuery, CHARACTER_ENCODING)
+                    );
+                    LOGGER.info(dateMetricsUrl);
+                    client.setBaseURL(dateMetricsUrl);
+                    JSONObject genericJsonObject = client.fetchMetricsJson(KruizeConstants.APIMessages.GET, "");
+                    JsonObject jsonObject = new Gson().fromJson(genericJsonObject.toString(), JsonObject.class);
+                    JsonArray resultArray = jsonObject.getAsJsonObject(KruizeConstants.JSONKeys.DATA).getAsJsonArray(KruizeConstants.DataSourceConstants.DataSourceQueryJSONKeys.RESULT);
```

Comment:
Could you please add a throw inside the getAsJsonArray method and propagate the exception to the caller, i.e., the GenerateRecommendation API?

Currently, if getAsJsonArray fails due to a 400, 500, or timeout error, the GenerateRecommendation API won't be notified of what happened, making it impossible to return a 400 status or a properly formatted error message to the API response.

Reply:
Added `fetchMetricsError` exception and propagating it to generate recommendations and added a custom message for the error with 400 error code.